File: mockHttp/src/test/java/project/steps/BattleShipSteps.java
package project.steps;

import org.testng.Assert;
import project.models.Field;
import project.models.TemplatesGame;
import project.pages.MainPage;

import java.util.ArrayList;
import java.util.LinkedHashSet;

public class BattleShipSteps {

    private static final int rangeForRandomPlaceShips = 15;

    public static void preparationToStart(MainPage mainPage) {
        mainPage.clickRandomOpponent();
        mainPage.clickRandomPlaceShipRandomTimes(rangeForRandomPlaceShips);
        mainPage.clickStartGame();
    }

    public static void playingGame(MainPage mainPage) {
        LinkedHashSet<Field> fields = new TemplatesGame().generateCombinationFieldsTemplate();
        for (Field f : fields) {
            if (mainPage.isInvisibleLabel()) {
                mainPage.clickField(f);
                if (mainPage.isInjureShip(f)) {
                    ArrayList<Field> nextInjuredFields = killShipFirstStage(f, mainPage);
                    if (!nextInjuredFields.isEmpty()) {
                        ArrayList<Field> finallyInjuredFields = killShipNextStage(nextInjuredFields, mainPage);
                        if (!finallyInjuredFields.isEmpty()) {
                            killShipNextStage(finallyInjuredFields, mainPage);
                        }
                    }
                }
            } else getResultGame(mainPage);
        }
    }

    private static ArrayList<Field> killShipFirstStage(Field injureField, MainPage mainPage) {
        ArrayList<Field> nextInjuredFields = new ArrayList<>();
        LinkedHashSet<Field> templateInjuredFields = new TemplatesGame().killShipTemplate(injureField);
        for (Field injuredField : templateInjuredFields) {
            mainPage.clickField(injuredField);
            if (mainPage.isInjureShip(injuredField)) {
                nextInjuredFields.add(injuredField);
            }
        }
        return nextInjuredFields;
    }

    private static ArrayList<Field> killShipNextStage(ArrayList<Field> injuredFields, MainPage mainPage) {
        ArrayList<Field> nextInjuredFields = new ArrayList<>();
        for (Field injuredField : injuredFields) {
            nextInjuredFields.addAll(killShipFirstStage(injuredField, mainPage));
        }
        return nextInjuredFields;
    }

    private static void getResultGame(MainPage mainPage) {
        Assert.assertTrue(mainPage.getAttributeFromLabelGame().contains("game-over-win"), mainPage.getTextFromLabelGame());
    }
}

File: ButtleShip/src/test/kotlin/steps/BattleShipSteps.kt
package steps

import framework.utils.Logger
import models.Field
import models.Templates
import org.testng.Assert
import pages.MainPage
import java.util.*

private val rangeForRandomPlaceShips = 15
private val logger: Logger = Logger()

fun preparationToStart() = with(MainPage()) {
    logger.info("Prepare to start")
    clickRandomOpponent()
    clickRandomPlaceShipRandomTimes(rangeForRandomPlaceShips)
    clickStartGame()
}

fun playingGame() = with(MainPage()) {
    logger.info("Playing game")
    Templates().generateCombinationFieldsTemplate()?.forEach() { field ->
        if (isVisibleLabel()) {
            getResultGame()
        } else {
            clickField(field!!)
            if (isInjureShip(field)) {
                killShipFirstStage(field).let { nextInjuredFields ->
                    if (nextInjuredFields.isNotEmpty()) {
                        killShipNextStage(nextInjuredFields).let { finallyInjuredFields ->
                            if (finallyInjuredFields.isEmpty()) {
                                killShipNextStage(finallyInjuredFields)
                            }
                        }
                    }
                }
            }
        }
    }
}

fun killShipFirstStage(injureField: Field?): ArrayList<Field> = with(MainPage()) {
    logger.info("Check fields after first kill field")
    val nextInjuredFields = ArrayList<Field>()
    Templates().killShipTemplate(injureField!!)?.forEach {
        clickField(it)
        if (isInjureShip(it)) {
            nextInjuredFields.add(it)
        }
    }
    return nextInjuredFields
}

private fun killShipNextStage(injuredFields: ArrayList<Field>): ArrayList<Field> = with(ArrayList<Field>()) {
    logger.info("Check fields after second and third kill field")
    injuredFields.forEach { this.addAll(killShipFirstStage(it)) }
    return this
}

fun getResultGame() = with(MainPage()) {
    logger.info("Get result game")
    Assert.assertTrue(
        getAttributeFromLabelGame()!!.contains("game-over-win"),
        getTextFromLabelGame()
    )
}